docker/docker_client.go

```diff
@@ -248,31 +259,52 @@ func newDockerClient(sys *types.SystemContext, registry, reference string) (*doc
 	// Check if TLS verification shall be skipped (default=false) which can
 	// either be specified in the sysregistriesv2 configuration or via the
 	// SystemContext, whereas the SystemContext is prioritized.
-	skipVerify := false
+	var sysContextSkipVerify *bool
 	if sys != nil && sys.DockerInsecureSkipTLSVerify != types.OptionalBoolUndefined {
 		// Only use the SystemContext if the actual value is defined.
-		skipVerify = sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
-	} else {
-		reg, err := sysregistriesv2.FindRegistry(sys, reference)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error loading registries")
+		skipVerify := sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
+		sysContextSkipVerify = &skipVerify
+	}
+
+	// Find the registry
+	reg, err := sysregistriesv2.FindRegistry(sys, ref)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+
+	// Create a fallback registry from the provided parameter
+	if reg == nil {
+		// The only source for InsecureSkipVerify can be the SystemContext
+		insecure := false
+		if sysContextSkipVerify != nil && *sysContextSkipVerify {
+			insecure = true
 		}
-		if reg != nil {
-			skipVerify = reg.Insecure
+
+		// Retrieve the docker client from the found registry
+		reg = &sysregistriesv2.Registry{
+			Endpoint: sysregistriesv2.Endpoint{
+				URL:      registry,
+				Insecure: insecure,
+			},
 		}
 	}
-	tr.TLSClientConfig.InsecureSkipVerify = skipVerify
 
+	// Rewrite the registry if we found the dockerHostname
+	if registry == dockerHostname {
+		reg.URL = dockerRegistry
```

Comment:
I think *this* special case belongs into `sysregistriesv2`. I don't like to post-process data from another API as it somehow shows a lack of functionality of the API. Maybe, `func parseURL(input string) (string, error)` can take care of that.

If we keep it as is, we rewrite what the user has configured which can lead to undesired behaviour, see @mtrmac's example above (i.e., `   URL: "docker.io/some-vendor/product"`).

Reply:
> I think _this_ special case belongs into `sysregistriesv2`. I don't like to post-process data from another API as it somehow shows a lack of functionality of the API.

I don’t quite view it that way; `sysregistriesv2` deals with _docker references_ (or prefixes, whatever), i.e. user-visible names, not URLs at the underlying transport level.  c/image/docker exists to translate those user-visible concepts into communication over the actual HTTP-based transport, and getting real data in/out.

Notably, consider the `SearchRegistry` case, where the user-visible `docker.io` registry name is translated into _two different_ host names (`index.docker.io`, `registry-1.docker.io`).  The way I think about it, `sysregistriesv2` conceptually does not care about v1/v2 API at all because the APIs are just not its concern — and it would need to know which API the caller intends to use to do this mapping.